### PR TITLE
refactor(_command_offset): save Bash's $2 passed to completion functions

### LINKED ISSUE #790: bash command completion has a problem with sudo command
## Describe the bug

When the command completion function starts running
there are three variables automatically setting `$1`, `$2`, `$3`.
and `$2` is not always the same as `${COMP_WORDS[COMP_CWORD]}`.

1. if `$2` value is a member of `$COMP_WORDBREAKS` variable
   then it's value is empty by default

   ex)  command --foo=[tab]
           `$2` : (empty)     `${COMP_WORDS[COMP_CWORD]}` : `=`

2. if `$2` value start with `"` or `'` (single or double quote)
   then it's quote removed ( `"foo` --> `foo` )

   ex)  command --foo "bar[tab]
           `$2` : `bar`         `${COMP_WORDS[COMP_CWORD]}` : `"bar`

this is an intended work for making completion function simpler
but if i use sudo in front of a command then
`$2` value is always the same as `${COMP_WORDS[COMP_CWORD]}`
This is a different behavior to the bash default and problematic.
and ends up can not use `$2` variable and force to use
`${COMP_WORDS[COMP_CWORD]}`  to support sudo

## Versions 

- [ ] Operating system name/distribution and version:
         Operating System: Ubuntu 22.04.1 LTS              
          Kernel: Linux 5.15.0-43-generic
         Architecture: x86-64
- [ ] bash version: 5.1.16(1)-release
- [ ] bash-completion version: 2.11

### PR DESCRIPTION
**Pull request by akinomyoga**
Fixes #790

This is a fix alternative to #791. Instead of imitating the arguments passed by Bash, we may save the original arguments in `_init_completion` (renamed to `_comp_initialize`). However, this involves not only the change of `_init_completion` itself but also the interface change of functions that call `_init_completion` internally---such as `_filedir_xspec`, `_longopt`, `_minimal`, `_known_hosts`, etc. Also, this doesn't work when `_comp_command_offset` is used without calling `_init_completion` (some cases are found in external completions). We need to request external completion writers carefully update their code.

### CODE REVIEW

File: bash_completion [hunk 1]
```diff
@@ -1010,6 +1024,8 @@ _init_completion()
                 ;;
         esac
     done
+    shift "$((OPTIND - 1))"
+    (($#)) && comp_args=("$@")
```
Comment (scop) [inline bash_completion:1028]: Would it make sense to do this only if `comp_args` is defined? Otherwise I suppose completions not prepared for the new behavior will leak a global by that name unless I'm missing something?

```suggestion
    if [[ -v comp_args ]]; then
        shift "$((OPTIND - 1))"
        (($#)) && comp_args=("$@")
    fi
```

File: bash_completion [hunk 2]
```diff
@@ -987,13 +987,27 @@ _comp_variable_assignments()
...
+#                     The first argument $1 is command name.  The second
+#                     argument $2 is the string before the cursor in the
+#                     current word.  The third argument $3 is the previous
+#                     word.
+# @var[out] cur       Reconstructed current word
+# @var[out] prev      Reconstructed previous word
+# @var[out] words     Reconstructed words
+# @var[out] cword     Current word index in `words`
+# @var[out] comp_args Original arguments specified to the completion function
+#                     are saved.
```
Comment (scop) [inline bash_completion:1000]:
```suggestion
#                     are saved in this array, if the variable is defined.
```

**Comment by scop** [comment]
Even though this PR is huge and there is an interface change involved, I feel the code here is much easier to understand than in #791; from that point of view I have a better feeling about this one.

If I understand correctly, the API change is such that it will mostly degrade gracefully, i.e. if the original arguments are not available, we'll basically do what we used to do before this change? Or will this flat out break some scenarios if not adjusted by affected completions?

External completion authors will likely be using `_init_completion` for a long time still. Luckily the current implementation (before this change) should be fine if it receives positional arguments, so I suppose externals can switch basically to
```bash
    local cur prev words cword split comp_args
    _init_completion -- "$@" || return
```
...in order to support earlier bash completion versions and to take advantage of the new stuff.

File: bash_completion [hunk 3]
```diff
@@ -987,13 +987,27 @@ _comp_variable_assignments()
...
+# @param $1...$3 args Original arguments specified to the completion function.
+#                     The first argument $1 is command name.  The second
+#                     argument $2 is the string before the cursor in the
+#                     current word.  The third argument $3 is the previous
+#                     word.
+# @var[out] cur       Reconstructed current word
+# @var[out] prev      Reconstructed previous word
+# @var[out] words     Reconstructed words
+# @var[out] cword     Current word index in `words`
+# @var[out] comp_args Original arguments specified to the completion function
```
Comment (scop) [inline bash_completion:999]: I have mixed feelings about this name: it's nice and concise, but it somehow slightly annoys me that the name does not make it crystal clear that these are the _original_ completion arguments. Dunno what to do about it if anything, not a huge fan of `comp_orig_args` (or `orig_args` or the same with `original` instead of `orig`)...

**Comment by akinomyoga** [comment]
> Even though this PR is huge and there is an interface change involved, I feel the code here is much easier to understand than in #791; from that point of view I have a better feeling about this one.

OK! Then, let us consider this direction!

> If I understand correctly, the API change is such that it will mostly degrade gracefully, i.e. if the original arguments are not available, we'll basically do what we used to do before this change?

Yes, it is basically right. Actually, it slightly improves the original behavior. Originally, we passed the entire word `${COMP_WORDS[COMP_CWORD]}` as the second argument, but now we are passing the part of the word after the current cursor position, i.e., `$cur`, as Bash does for its built-in programmable completion.

> External completion authors will likely be using `_init_completion` for a long time still. Luckily the current implementation (before this change) should be fine if it receives positional arguments, so I suppose externals can switch basically to
> 
> ```shell
>     local cur prev words cword split comp_args
>     _init_completion -- "$@" || return
> ```
> 
> ...in order to support earlier bash completion versions and to take advantage of the new stuff.

Ah, this is a good point. I hadn't noticed that point by myself. Thank you.

File: bash_completion (hunk 1 above)
Comment (akinomyoga) [reply bash_completion:1028]: `local ... comp_arg` doesn't make the variable *set state*. For `[[ -v comp_args ]]` to success, an actual value should be assigned to `comp_arg` prior to the test. So,

1. one way is to request writing like `local ... comp_arg=`.
2. Another way is to test the existence of the variable by `declare comp_args &>/dev/null`.

But I'm not sure if we need to test it explicitly because I think we can assume that `comp_args` is declared when extra arguments are specified to `_comp_initialize` or `_init_completion`. This is tested by `(($#)) &&`.

File: bash_completion (hunk 3 above)
Comment (akinomyoga) [reply bash_completion:999]: It was actually the same for me. I also thought of different names such as `original_args`, `comp_original_args`, `old_args`, `args`, etc., but some seemed too long, and the other seemed too general. I thought `prev` and `words` are already general names so initially thought of using `args`, but it turned out `args` are already widely used by completion functions for other purposes. If we wouldn't use `comp_args`, my second candidate is `original_args`.

Comment (scop) [reply bash_completion:1046]: Ok, let's go with `comp_args`.

**Review by scop — APPROVED** [review]
Ok, let's go with this. Thanks!
